Implement TarOperation and TarParam foundations

stillbeingnick opened the pull request:
This PR builds on the great work done by @kaladron and lays out the foundation to enable adding operations and options moving forward. Leveraging rust enums and building behavioral control with the enum dispatch pattern.

There are 3 critical pieces to this foundational layer.

### TarParams
- TarParams provides highline access to arguments passed in and some convenience to the options set by the user that are called during execution. This also serves as the functional place that clap arguments are processed from uu_main. ArgMatches are pumped in and TarParams, TarOptions, and the TarOperation pop out.
### TarOptions
- An enum used to capture arguments and potential subarguments passed in via clap and command line. Due to the overwhelming amount of options that tar has, we needed a way to store these while also potentially caputuring data that could be associated with that option. Like BlockingFactor might take in a u32, while Verbose is only a flag so its presentance in a list is enough, and a rust enum is the best way to handle this wide swath of possibilities. A Vec of these options is stored in TarParams and created when TarParams::with_operation() is called.
### TarOperation
- This trait serves as the execution hook from processing from clap ArgMatchees to execution. Each Main Operation mode is created as a file (create.rs, extract.rs) and each of these have a new type associated with their respected operation. Which then impl the TarOperation trait and the exec method. This gives us a place to call our main operation modes and decouple each operation from one another, but also allowing the reuse of operations in other operations. For example, using append in the update operation.

While this is just the bones it will continue to morph and change as more functionality is built into tar.

stillbeingnick committed:
Adding TarParam and TarOperation framework

clippy and cargo fmt

stillbeingnick committed:
Adding TarParam and TarOperation framework

clippy and cargo fmt

stillbeingnick committed:
Adding TarParam and TarOperation framework

clippy and cargo fmt

stillbeingnick committed:
Adding TarParam and TarOperation framework

clippy and cargo fmt
changing arg match taroption processing

Removing for match loop going over all matched args. Moved to standard if get_* for specific matches.

fmt and derive default tarparams

cargo workspace dependency fixes
Adding TarParam and TarOperation framework

clippy and cargo fmt
changing arg match taroption processing

Removing for match loop going over all matched args. Moved to standard if get_* for specific matches.
Revert "changing arg match taroption processing"

This reverts commit 4bef208161a1271385119982c29aa3d399ed90ff.

changing arg match taroption processing

Removing for match loop going over all matched args. Moved to standard if get_* for specific matches.
cargo workspace dependency fixes
Adding TarParam and TarOperation framework

clippy and cargo fmt
changing arg match taroption processing

Removing for match loop going over all matched args. Moved to standard if get_* for specific matches.

fmt and derive default tarparams

stillbeingnick commented:
Wanted to give this another bump @sylvestre @cakebaker. The PR after this I plan to rework the Clap arg processing so we can not only take care of @sylvestre suggestion to help make TarOperation more sane but also see if we can start addressing #55.

stillbeingnick committed:
changing arg match taroption processing

Removing for match loop going over all matched args. Moved to standard if get_* for specific matches.

stillbeingnick committed:
changing arg match taroption processing

Removing for match loop going over all matched args. Moved to standard if get_* for specific matches.

stillbeingnick committed:
changing arg match taroption processing

Removing for match loop going over all matched args. Moved to standard if get_* for specific matches.

stillbeingnick commented on src/uu/tar/src/tar.rs:
```diff
@@ -33,45 +36,10 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
-    ))
+    op.exec(&options)
 }
-
```
Added space between `uu_main()` and `uu_app`. Re-ran cargo fmt for good measure.

stillbeingnick committed:
fmt and derive default tarparams

stillbeingnick committed:
fmt and derive default tarparams

stillbeingnick committed:
Merge branch 'main' into feat/taroperation-tarparam

stillbeingnick committed:
merge for create verbose

sylvestre commented on src/uu/tar/Cargo.toml:
```diff
@@ -17,6 +17,7 @@ uucore = { workspace = true }
 clap = { workspace = true }
 regex = { workspace = true }
 tar = { workspace = true }
+jiff = "0.2.16"
```
please use the workspace style declaration

stillbeingnick committed:
cargo workspace dependency fixes

stillbeingnick committed:
cargo workspace dependency fixes

stillbeingnick committed:
cargo workspace dependency fixes
Adding TarParam and TarOperation framework

clippy and cargo fmt
changing arg match taroption processing

Removing for match loop going over all matched args. Moved to standard if get_* for specific matches.

fmt and derive default tarparams

stillbeingnick committed:
Merge local to origin

stillbeingnick committed:
Revert "changing arg match taroption processing"

This reverts commit 4bef208161a1271385119982c29aa3d399ed90ff.

stillbeingnick committed:
Merge remote-tracking branch 'origin/feat/taroperation-tarparam' into feat/taroperation-tarparam

stillbeingnick committed:
Merge remote-tracking branch 'origin/feat/taroperation-tarparam' into feat/taroperation-tarparam

sylvestre commented on src/uu/tar/src/options/options.rs:
```diff
@@ -0,0 +1,92 @@
+/// [`TarParams`] Holds common information that is parsed from
+/// command line arguments. That changes the current execution of
+/// tar.
+#[allow(dead_code)]
```
can we avoid this ?
it is usually a sign that the code needs to be refactored

stillbeingnick committed:
remove dead and detail tarparam error message

stillbeingnick committed:
Merge branch 'main' into feat/taroperation-tarparam
